internal/docs/design.md

```diff
@@ -374,6 +375,100 @@ b, err := blob.NewBucket(d, blob.FeatureUnicodeNames)
 Design discussions regarding enforcing portability are ongoing; we welcome input
 on the [mailing list](https://groups.google.com/forum/#!forum/go-cloud).
 
+### Strings
+
+Providers often differ on what characters they accept in particular strings
+(e.g., blob names, metadata keys, etc.). A couple of specific examples:
+
+*   Azure Blob only
+    [accepts C# identifiers](https://docs.microsoft.com/en-us/azure/storage/blobs/storage-properties-metadata)
+    as metadata keys.
+*   S3 drops double slashes in blob names (e.g., `foo//bar` will end up being
+    saved as `foo/bar`).
+
+These differences lead to a loss of portability and predictability for users.
+
+We considered restricting Go CDK's APIs to strings that all providers support.
+For example, we asserting that Go CDK's blob only supports ASCII plus `/` for
+blob names (and no `//`!). However, such a rule would mean that we couldn't
+handle existing strings that were created through some mechanism other than
+through Go CDK APIs. For example, an existing blob in S3 with a unicode name
+wouldn't be visible in Go CDK. Or maybe it would be visible, but you wouldn't be
+able to write to it. For metadata keys, rules like this might mean that you
+could read metadata attributes via the Go CDK and then write them back, and
+either lose data (if previously-existing invalid keys were dropped) or get an
+error. Overall, we decided that this approach is unacceptable.
+
+Instead, we assert that Go CDK can handle any string, and force provider
+implementations to handle strings that the underlying provider can't handle
+using escaping mechanisms. We enforce provider compliance with conformance
+tests. With this approach
+
+*   Go CDK APIs have visibility into all existing strings for all providers.
+    *   A slight caveat -- strings that look like Go CDK-escaped strings will be
+        unescaped by Go CDK. To mitigate this, we'll choose unusual-looking
+        escape mechanisms.
+*   Go CDK APIs are consistent in that you can write any string to any provider
+    and read it back.
+*   Some strings that were written through the Go CDK will appear in their
+    escaped form when viewed outside of Go CDK.
+
+The Go CDK providers helpers for such escaping:
+
+```
+package stringescape
```

Comment:
Maybe just `escape`? That's a long package name.

I propose we put this and similar packages in a `support` subdir to avoid cluttering the root dir.

Reply:
I don't see a need to export this as a whole package yet. Especially given the rollout plan of "wait until need demands to expose", I would rather the escaping start as fixed functions in each of the driver implementations, and then export a pattern once we find it.

Reply:
@zombiezen -- why? I'm intending to use the exact same escaping code for every provider. Because, why not? They differ in which bytes need escaping, not how to do the escaping.

I was planning on putting this code in `internal/`, so it's not really exported outside, but usable from multiple providers.

Reply:
Summarizing offline remarks: IMO the prose would be easier to read in this order:

1.  Brief problem statement
2.  Normative declaration of how we want this problem solved (all the immutable principles)
3.  Non-normative example of how to solve it for a particular set of constraints (e.g. C# identifiers)
4.  Rationale/Alternatives Considered